src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -374,6 +377,11 @@ public void Reset()
 
             _responseBytesWritten = 0;
             _requestCount++;
+
+            // TODO: probably not the right place to complete
+            RequestBodyPipe.Reader.Complete();
+            RequestBodyPipe.Writer.Complete();
+            RequestBodyPipe.Reset();
```

Comment:
The call to Reset can be skipped prior to the first request. You could even move the call to `RequestBodyPipe.Reset()` out of `Frame.Reset()` and directly into the end of `ProcessRequestsAsync()` loop if that's cleaner.